private/buf/bufformat/testdata/proto2/message/v1/message_options.proto

```diff
@@ -13,8 +13,30 @@ option (custom.message_thing_option) = {
 
  /* Leading comment on the 'foo' element */ foo:  1  /* Comment on bar */ bar:   2  /* Comment on truth */  truth:false    }  ;
 
+
+
+      option (custom.message_thing_option)=  { /*leading*/ foo: 1 /*trailing*/ } /*leading*/;/*trailing*/
```

Comment:
@pkwarren, I updated the formatter like you requested, to omit whitespace after open parens and braces and before close parens and braces.

I then decided to overhaul this file with a bunch of new examples with various kinds of braces/brackets/parens, to make sure the output looked good. And I discovered a few issues.

Unfortunately, the fixes to these issues _will_ produce some diff churn for users .But I think it is worth it because the new canonical format is _much_ easier to read IMO.

Here are the things I fixed:
* If an array literal or message literal had an inline comment before the closing brace, it was printed in a rather verbose and bad-looking way:
   ```
   [
      123

   /* inline comment */
   ]
   ```
   So I've fixed this to look better IMO:
   ```
   [
      123
      /* inline comment */
   ]
   ```
* There was an exception for simple compact options to be printed inline, on a single line. But there was no such exception for simple (single element) arrays and messages. So I've added one. If the array or message has a single element/field whose value is not a nested array or element, it can be printed all on one line. You can see an example in field `d` of the complex option below.

Reply:
Welp there are some nasty issues in this code. One is that it interprets "trailing comment" to mean something a little different than what it is. And, honestly, this probably warrants a change in protocompile.

Currently, the lexer attempts to classify comments as "trailing" for previous token or "leading" for current/next token. It does this using heuristics that try to emulate how `protoc` does this, for purposes of matching its source code info in descriptors.

_However_, those heuristics are too complicated and cause the interpretation of a comment as leading vs trailing to vary with subtle newline changes. That means that the act of formatting the code like above (removing an extra newline for compact options, message literals, and array literals) and removing a newline can cause a comment to be interpreted in a re-format as a trailing comment (when it was previously interpreted as a leading comment). That means that the format is inherently unstable: reformatting in that case would change it further.

I think what we really want is that the AST only calls something a trailing comment if it is on the same line as the element. And then, when computing source code info for the AST, the compiler may have to "borrow" some leading comments from the next token to populate trailing comments like `protoc` does. This "borrow" step being in the lexer and impacting the AST is the root issue right now.

Also, this code is very hard to maintain. There are are _lots_ of helper functions that are not named what they are ostensibly _for_ instead of being named after what they actually do. For example, there is `writeBodyEnd`, which is used to write the end of a body... but also used in myriad other places, such as elements in a compound string literal 🤷.

The way the helper methods are shared/re-used, but without clear names, means its easy to make a seemingly reasonable change in a helper but it to have surprising side effects because of how its re-used in weird places. Anyhow, I feel like I've only made the code messier in trying to get this to look a little better. And now I know just how brittle it is.

I think I will put this PR on hold and go and make the above-described changes to protocompile -- so the AST is more predictable and simple with what gets classified as a trailing comment and move the logic to classify like `protoc` does into its `sourceinfo` package. That will be a much more stable point on which to implement formatter improvements.